### PR TITLE
Add an initial proposal for the Flush RPC.

### PR DESCRIPTION
**Pull request by robshakir**
```
 * (M) v1/proto/service/gribi.proto
  - Add a Flush RPC proposal.
```

Proposal for the Flush RPC. The operation supports modes whereby the client expects that it is the master, and where it does not know the current election ID. Flushes are provided per network instance, or for all network instances.

### CODE REVIEW

File: v1/proto/service/gribi.proto [hunk 1]
```diff
@@ -421,3 +427,49 @@ message GetResponse {
...
+    bool override = 2;
+  }
+  // network_instance specifies the Network Instance that should be flushed.
+  oneof network_instance {
+    // The network instance from which the entries should be retrieved.
+    // If name is specified the empty string "" is considered invalid.
+    string name = 3;
+    // all indicates that entries for all network instances for the
+    // specified network instance should be returned.
+    Empty all = 4;
```
Comment (nandanarista) [inline v1/proto/service/gribi.proto:452]: copy paste error on the comments (`should be retrieved`/ `returned` needs to be `flushed`)

File: v1/proto/service/gribi.proto [hunk 2]
```diff
@@ -421,3 +427,49 @@ message GetResponse {
   // A set of entries that are installed on the device.
   repeated AFTEntry entry = 1;
 }
+
+// FlushRequest is the request sent from the client to the server when a Flush
+// RPC is called.
+message FlushRequest {
+  // election specifies the election ID from a client. It is valid only
```
Comment (nandanarista) [inline v1/proto/service/gribi.proto:434]: if specified in ALL_PRIMARY should `election` be ignored by the server ? or should it reject the request

File: v1/proto/service/gribi.proto [hunk 3]
```diff
@@ -421,3 +427,49 @@ message GetResponse {
...
+  // when the server has clients that are operating in the SINGLE_PRIMARY mode.
+  oneof election {
+    // id indicates that the Flush ID supplied should be compared to the
+    // existing election ID on the server from the current master.
```
Comment (nandanarista) [inline v1/proto/service/gribi.proto:438]: should we add more text here. I'm guessing the `id` needs to be equal to the primary's election id for the server to accept the request. What if it's > than the primary's? Is that valid?

Note that comparing against current primary means that if the primary is disconnected or there are no ModifyRPC streams, even though the server knows the previous highest election_id it'll reject the flush request.
In other words, flush will be honored if id == primary's election_id and primary is connected.

File: v1/proto/service/gribi.proto (hunk 1 above)
Comment (robshakir) [reply v1/proto/service/gribi.proto:452]: Thanks, fixed.

File: v1/proto/service/gribi.proto (hunk 3 above)
Comment (robshakir) [reply v1/proto/service/gribi.proto:438]: Thanks for catching this -- yes, definitely agree that there is some work here to be clearer, how does the following text sound: 

```
    // id indicates that the Flush ID supplied should be compared to the
    // existing election ID on the server. The logic
    // for whether the Flush RPC is accepted is the same as that which is
    // implemented for Modify. In summary:
    //
    //   - if the election ID is equal to the previous highest known ID, the Flush
    //     is accepted.
    //   - if the election ID is greater than the previous highest known ID, the
    //     Flush is accepted.
    //   - if the election ID is less than the previous highest known ID, the
    //     Flush is rejected and an error with reason NOT_MASTER is returned.
    //
    // If the server is currently operating in ALL_PRIMARY mode the server should
    // return an error specifying ELECTION_ID_IN_ALL_PRIMARY reason.
```

File: v1/proto/service/gribi.proto [hunk 4]
```diff
@@ -421,3 +427,64 @@ message GetResponse {
...
+    // clients on the server.
+    bool override = 2;
+  }
+  // network_instance specifies the Network Instance that should be flushed.
+  oneof network_instance {
+    // The network instance from which the entries should be flushed.
+    // If name is specified the empty string "" is considered invalid.
+    string name = 3;
+    // all indicates that entries for all network instances for the
+    // specified network instance should be flushed.
```
Comment (nandanarista) [inline v1/proto/service/gribi.proto:463]: There's an extraneous `for the specified network instance`.

```
all indicates that entries for all network instances should be flushed.
```

**Review by nandanarista — APPROVED** [review]
LGTM based on the comment about the intent of Flush RPC.
I have just one open comment to fix the comment for `all` network_instance.

**Review by nflath — APPROVED** [review]

File: v1/proto/service/gribi.proto (hunk 4 above)
Comment (robshakir) [reply v1/proto/service/gribi.proto]: Thanks, this was both a copy-pasta and an error in `Get`. Fixed.
